docs/proposals-accepted/202304-cardinalitymanagment.md

```diff
@@ -0,0 +1,136 @@
+- https://mentorship.lfx.linuxfoundation.org/project/dbce5279-d029-46f3-b117-9e9dd7f84bd6
+
+### Why
+
+At present, multiple product teams send their metrics to Thanos. However, when product teams send metrics with high cardinality to Thanos, it can overwhelm the Store Gateway and Queriers, preventing any teams from running queries. To prevent this issue from occurring, we should have a cardinality management support in Thanos. This would allow teams to monitor the cardinality of both labels and metrics.
+
+### Pitfalls of current solutions
+
+Currently, Thanos receiver supports the TSDB stats API on remote write addresses, which provides information about series cardinality. However, this API can only return a limited number of results per query, typically no more than 10. This limitation makes it difficult to track arbitrary metrics and may lead to incomplete or inaccurate cardinality data. To address this issue, there should be some Top-K selector that can return the stats of the top K entries, allowing teams to more effectively monitor cardinality and identify potential issues.
+
+### Goals
+
+Implement APIs exposing series cardinalities in Thanos.
+
+### Proposal
+
+We will be Developing a New API in Thanos to calculate series cardinality through Thanos Receiver.
+
+#### API Design
+
+The new API will have the following endpoint:
+
+`/api/v1/cardinality`: It will retrieve the series cardinality information.
+The API response will contain
+
+- `totalSeries`: Total number of series.
+- `totalLabelValuePairs`: Total number of label-value pairs.
+- `seriesCountByMetricName`: List of objects containing metric name and its corresponding series count.
+- `seriesCountByLabelName`: List of objects containing label name and its corresponding series count.
+- `seriesCountByFocusLabelValue`: List of objects containing focus label-value and its corresponding series count.
+- `seriesCountByLabelValuePair`: List of objects containing label-value pair and its corresponding series count.
+- `labelValueCountByLabelName`: List of objects containing label name and its corresponding label value count.\*
+
+The API response will be 
+
+```{
+  "status": "success",
+  "isPartial": false,
+  "data": {
+    "totalSeries": 14752,
+    "totalLabelValuePairs": 100544,
+    ... // other data fields
+  }
+}
+```
+
+#### Implementation
+
+For now extending receiver, seems the best approach to use. And, exact implementation can change while implementing the proposed API design.
+
+1. We will first access and aggregate the series, labels, and label-value pairs information within the Thanos receiver by extending its functionality while processing incoming data. The Thanos receiver ingests data and writes it to a TSDB instance.
+2. There will be a new data structure to hold the statistics. This structure should have methods to
+
+   - Add new data points (series, labels, and label-value pairs).
+   - Compute the statistics required by the API.
```

Comment:
I think this is the most important part. Let's not skip the details on how we would compute it.

Reply:
Yeah